### DISCUSSION TITLE
Local mqtt

### DISCUSSION THREAD

**Discussion by jorik41** [tolwi/hassio-ecoflow-cloud#88] [Ideas]
I'm inquiring about the possibility and complexity involved in setting up a local MQTT server. My goal is to route all the traffic aimed at mqtt.ecoflow.com to my own MQTT server on my local network. However, I'm unsure if this approach is even feasible given the potential complexities and technical challenges. I would appreciate any insights or guidance on this matter, especially from those who might have experience or expertise in similar setups. Can this be done and, if so, what would be required to accomplish it?

**Comment by tolwi** [comment]
https://github.com/mmiller7/ecoflow-withoutflow/blob/main/cloud-mqtt/ecoflow_get_mqtt_login.sh#L165 - i use such bridge to analyze messages in a local Mosquitto broker

**Reply by Ne0-Hack3r** [reply]
That is bridging the MQTT data from the EF cloud server to a locally maintained broker. I had done this as well and used my own YAML packages to create sensors in Home Assistant. However, in that setup the device is still subscribing to the cloud server and the local broker must also subscribe to that same cloud server (thus cloud dependency and all telemetry is provided to Ecoflow).

I think OP is asking about the possibility of intercepting the cloud communication to keep all MQTT local. I toyed with this concept using a DNS intercept but did not get too far. I have discovered, however, that for certain devices (like D2 and D2M) a BLE command exists to instruct the device to connect to WiFi specifying the URL of the certification server. The device then connects to that server and is provided JSON with another URL and credentials for MQTT. Using that command it is possible to setup an HTTP intercept ("man in the middle") which alters the JSON to tell the device to connect to a local MQTT server. Properly executed, the device then subscribes to my local Mosquito instance on Home Assistant and the EF cloud server is no longer required. Data updates and setting changes are much more responsive in this mode.

Alas, this does not work for SHP or DP (which appear to use byte stream communication on the back end over a hard coded port and URL). DP can be locally controlled over TCP or BLE using byte stream, which is what the https://github.com/vwt12eh8/hassio-ecoflow integration exploits (over TCP). At this point, I have local control of everything but my Smart Home Panel and have, thus far, found no way to open TCP communication. BLE is possible but requires manually turning BLE on using IoT button on SHP and would require a BT relay of some kind to keep that connection open and bridge to my Home Assistant VM.
